[BUG] preinstall script with npx does not work with --package-lock-only

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

A preinstall script using `npx` fails when master command `npm install --package-lock-only` is fired.
It works when master command is `npm install`

Behaviors differs between npm 6.x & npm 8.x but none are correct.


### Expected Behavior

`npm install --package-lock-only` shoud be successfull and should execute correctly the presinstall script.

### Steps To Reproduce

- `mkdir reproducer && cd reproducer`
- `nvm use 16`
  - at the time of filling bug it installs : `node v16.14.0 (npm v8.3.1)`
- `rm -rf ~/.npm/_npx` using a clean npx cache
- using the following package
```
cat <<EOF > package.json
{
    "name": "issue-only-allow-npx-package-lock-only",
    "version": "1.0.0",
    "description": "'npm install' works, 'npm install --package-lock-only' fails",
    "scripts": {
        "preinstall": "npm_config_yes=true npx only-allow npm"
    }
}
EOF
```
- `npm install --package-lock-only` fails

```
npm install --package-lock-only                                                      

> issue-only-allow-npx-package-lock-only@1.0.0 preinstall
> npm_config_yes=true npx only-allow npm

sh: 1: only-allow: not found
npm ERR! code 127
npm ERR! path /home/m.brouillard/projects/lectra/ui/issue-only-allow-npx-only-allow
npm ERR! command failed
npm ERR! command sh -c npm_config_yes=true npx only-allow npm

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/m.brouillard/.npm/_logs/2022-03-22T08_24_16_404Z-debug-0.log
```


### Environment

- npm: 8.3.1
- Node.js: 16.14
- OS Name: Ubuntu 18.04
- System Model Name: HP Probook

--- a/lib/commands/exec.js
+++ b/lib/commands/exec.js
@@ -73,6 +73,9 @@ class Exec extends BaseCommand {
 
     return libexec({
       ...flatOptions,
+      // we explicitly set packageLockOnly to false because if it's true
+      // when we try to install a missing package, we won't actually install it
+      packageLockOnly: false,
       args,
       call,
       localBin,

--- a/test/lib/commands/exec.js
+++ b/test/lib/commands/exec.js
@@ -385,6 +385,64 @@ t.test('npm exec foo, not present locally or in central loc', async t => {
   ])
 })
 
+t.test('npm exec foo, packageLockOnly set to true', async t => {
+  const path = t.testdir()
+  const installDir = resolve('npx-cache-dir/f7fbba6e0636f890')
+  npm.localPrefix = path
+  npm.config.set('package-lock-only', true)
+  t.teardown(() => {
+    npm.config.set('package-lock-only', false)
+  })
+
+  ARB_ACTUAL_TREE[path] = {
+    inventory: {
+      query () {
+        return new Set()
+      },
+    },
+  }
+  ARB_ACTUAL_TREE[installDir] = {
+    inventory: {
+      query () {
+        return new Set()
+      },
+    },
+  }
+  MANIFESTS.foo = {
+    name: 'foo',
+    version: '1.2.3',
+    bin: {
+      foo: 'foo',
+    },
+    _from: 'foo@',
+  }
+  await exec.exec(['foo', 'one arg', 'two arg'])
+  t.strictSame(MKDIRPS, [installDir], 'need to make install dir')
+  t.match(ARB_CTOR, [{
+    path,
+    packageLockOnly: false,
+  }])
+  t.match(ARB_REIFY, [{
+    add: ['foo@'],
+    legacyPeerDeps: false,
+    packageLockOnly: false,
+  }], 'need to install foo@')
+  t.equal(PROGRESS_ENABLED, true, 'progress re-enabled')
+  const PATH = `${resolve(installDir, 'node_modules', '.bin')}${delimiter}${process.env.PATH}`
+  t.match(RUN_SCRIPTS, [
+    {
+      pkg: { scripts: { npx: 'foo' } },
+      args: ['one arg', 'two arg'],
+      banner: false,
+      path: process.cwd(),
+      stdioString: true,
+      event: 'npx',
+      env: { PATH },
+      stdio: 'inherit',
+    },
+  ])
+})
+
 t.test('npm exec foo, not present locally but in central loc', async t => {
   const path = t.testdir()
   const installDir = resolve('npx-cache-dir/f7fbba6e0636f890')

